src/ir/auth_logic/ast_construction.cc

```diff
@@ -0,0 +1,164 @@
+//-------------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//-------------------------------------------------------------------------------
+
+#include <string>
+#include <vector>
+#include <utility>
+#include <stdexcept>
+#include <iostream>
+#include <memory>
+#include <variant>
+#include <optional>
+
+
+#include "antlr4-runtime.h"
+#include "absl/status/statusor.h"
+#include "src/common/utils/map_iter.h"
+#include "src/ir/auth_logic/ast.h"
+#include "src/ir/auth_logic/ast_construction.h"
+#include "src/policy/auth_logic/auth_logic_cc_generator_grammar.inc/AuthLogicLexer.h" 
+#include "src/policy/auth_logic/auth_logic_cc_generator_grammar.inc/AuthLogicParser.h"
+#include "src/policy/auth_logic/auth_logic_cc_generator_grammar.inc/AuthLogicVisitor.h"
+#include "src/ir/datalog/program.h"
+
+namespace raksha::ir::auth_logic {
+
+Principal Parser::ConstructPrincipal (AuthLogicParser::PrincipalContext* context) {
+    std::string name = context->ID()->getText();
```

Comment:
I see few problem when using const reference instead of pointers. Please let me know if you have a way to deal with this.
Original code:
```template<typename T, typename = std::enable_if_t<std::is_base_of<AuthLogicParser::FactContext, T>::value>>
T* GetFactContextAs(AuthLogicParser::FactContext* context) {
     return dynamic_cast<T*>(context);
}

absl::StatusOr<Fact> Parser::ConstructFact(AuthLogicParser::FactContext* context) {
    if (auto* flat_fact_fact_ctx = GetFactContextAs<AuthLogicParser::FlatFactFactContext>(context)) {  
        absl::StatusOr<BaseFact> flatfact_or_null = Parser::ConstructFlatFact(flat_fact_fact_ctx->flatFact());
        if (!flatfact_or_null.ok()) return absl::InvalidArgumentError("Invalid context to ConstructFact");
        BaseFact basefact = *flatfact_or_null;                               //FlatFactFactContext
        return Fact(basefact);
    }
    else if (auto* ctx = GetFactContextAs<AuthLogicParser::CanSayFactContext>(context))  {                                                      
        return Parser::ConstructFact(ctx->fact());
    }

    return absl::InvalidArgumentError("Invalid context to ConstructFact");
}```
Modified code with above suggestions:
```template<typename T, typename = std::enable_if_t<std::is_base_of<AuthLogicParser::FactContext, T>::value>>
T GetFactContextAs(AuthLogicParser::FactContext &context) {
     return dynamic_cast<T&>(context);
}

Fact Parser::ConstructFact(AuthLogicParser::FactContext &context) {
    if (auto flat_fact_fact_ctx = GetFactContextAs<AuthLogicParser::FlatFactFactContext>(context)) {  
        BaseFact basefact = Parser::ConstructFlatFact(flat_fact_fact_ctx->flatFact());
        return Fact(basefact);
    }
  else {
    auto ctx = GetFactContextAs<AuthLogicParser::CanSayFactContext>(context))                                                       
    return Parser::ConstructFact(ctx->fact());
}
 
}```

In the original code, dynamic cast on pointer, if the cast fails, it would return a NULL. This lets us know if the object is of that particular type or not( In our case, flatFactFactContext). If it is not, we check the same for another type (CanSayFactContext).
When casting a reference, it is not possible to return a NULL pointer to indicate failure; a dynamic cast of a reference variable will throw the exception std::bad_cast. This causes a problem as we never get to the else part.

Reply:
Taking a pointer instead of a reference in checking whether the result of the dynamic cast is `NULL` or not is fine. Cases like the above are exempt from the guidance I offered, as we expect `NULL` to be returned by `dynamic_cast` in the normal running of the program. In general, though, if using a reference type can prevent you from having to check whether a pointer is `NULL`, and all you would do if that pointer was `NULL` would be issue an unrecoverable error, it's better to use the reference type.